src/plugins/intel_gpu/src/graph/impls/common/loop.cpp

```diff
@@ -37,63 +107,169 @@ struct loop_impl : typed_primitive_impl<loop> {
     void set_node_params(const program_node& arg) override {
         OPENVINO_ASSERT(arg.is_type<loop>());
         const auto& node = arg.as<loop>();
-        _max_iteration = node.get_max_iteration();
         _back_edges = node.get_back_edges();
     }
 
+    void set_memory_in_body_network(cldnn::network::ptr body_network,
+                    const std::shared_ptr<cldnn::primitive_inst>& inst, memory::ptr mem) const {
+        if (inst->is_input()) {
+            body_network->set_input_data(inst->id(), mem);
+        } else if (inst->is_output()) {
+            body_network->set_output_memory(inst->id(), mem);
+        } else {
+            inst->set_output_memory(mem, false);
+        }
+    }
+
+    std::vector<event::ptr> handle_buffers_for_next_iteration(const loop_inst::backedge_memory_mapping& mapping,
+                                                    network::ptr body_network, int64_t iter, bool is_dynamic) const {
+        std::vector<event::ptr> event_vec;
+        OPENVINO_ASSERT(iter >= 0, "iteration should not be negative : ", iter);
+        if (mapping.type == loop_inst::backedge_memory_mapping::CONCAT_OUTPUT) {
+            if (iter == 0) {
+                set_memory_in_body_network(body_network, mapping.to_primitive, mapping.initial_mem);
+            } else if (iter > 0) {
+                if (is_dynamic) {
+                    auto from_id = mapping.from_primitive->id();
+                    if (body_network->has_event(from_id)) {
+                        auto ev = body_network->get_primitive_event(from_id);
+                        if (ev) ev->wait();
+                    }
+                    // In dynamic model, just copy data from inner body output to inner body input in back_edges.
+                    memory::ptr mem1 = mapping.to_primitive->output_memory_ptr();
+                    memory::ptr mem2 = mapping.from_primitive->output_memory_ptr();
+                    auto ev = mem1->copy_from(body_network->get_stream(), *(mem2));
```

Comment:
By default copy_from() call is blocking, so you can ignore output event or need to call this function with explicit blocking = false